xwiki-rendering-macros/xwiki-rendering-macro-message/src/main/java/org/xwiki/rendering/internal/macro/message/AbstractMessageMacro.java

```diff
@@ -73,4 +79,29 @@ public void prepare(MacroBlock macroBlock) throws MacroPreparationException
     {
         this.contentParser.prepareContentWiki(macroBlock);
     }
+    
+    protected ResourceReference getIconReference()
+    {
+        IconResourceReferenceTypeParser iconParser = new IconResourceReferenceTypeParser();
+        return iconParser.parse(this.iconName);
+    }
+
+    @Override
+    public List<Block> execute(BoxMacroParameters parameters, String content, MacroTransformationContext context) 
+        throws MacroExecutionException 
+    {
+        List<Block> boxFundation = super.execute(parameters, content, context);
+        if (boxFundation.size() == 0 || this.iconName == null) {
+            return boxFundation;
+        } else {
+            // Enhance the default box with an icon in the top left.
+            Block defaultBox = boxFundation.get(0);
+            Block iconBlock = new ImageBlock(this.getIconReference(), true);
```

Comment:
An actual image inserted in the content does not feel like the cleanest. I would have expected the exact icon to be a CSS (so on platform side) decision based on the class (so, like the color).

Reply:
We want the icon to be accessible from a screen reader, so unfortunately we can't just add it in the `:before` pseudo-element of the message box.

I thought the best practice would be to add an image in it, but I think I could also just add an empty `Format Block` just like for the title, and somehow fill it with CSS.

However AFAIK, the various iconthemes also have various ways to represent icons in HTML, some use images, some use fonts, ... so I'm not sure CSS only would solve this completely.

Reply:
We have unfortunately no way to insert icons by CSS unless I missed something. Regarding the screen reader aspect, that's easy to solve I think by adding a `sr-only` text. But then again, I'm not really sure that this is something we should do in rendering as `sr-only` is not a standard but comes from Bootstrap iirc. Imho it would be cleanest to override these macros in `xwiki-platform` to add proper icon HTML from the icon theme.